### PR TITLE
[Model] support minicpm3

### LINKED ISSUE #8232: [New Model]: when support MiniCPM3ForCausalLM  MiniCPM3-4B model
### The model to consider.

https://huggingface.co/openbmb/MiniCPM3-4B

### The closest model vllm already supports.

_No response_

### What's your difficulty of supporting the model you want?

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by SUDA-HLT-ywfang**
Support MiniCPM3-4B.
github repo: https://github.com/OpenBMB/MiniCPM
huggingface repo: https://huggingface.co/openbmb/MiniCPM3-4B

FIX #8232
**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Thanks for implementing this! It looks like there are quite a few duplicate modules with the old MiniCPM though. If you don't plan to modify those modules, could you import the existing ones from `minicpm.py`?

**Comment by DarkLight1337** [comment]
Also, it would be great if you could add a test to show that the model matches the HF output.

File: vllm/model_executor/models/minicpm3.py [hunk 1]
```diff
@@ -304,22 +181,22 @@ def forward(
...
+class MiniCPM3DecoderLayer(MiniCPMDecoderLayer):
 
     def __init__(
         self,
         config,
         cache_config: Optional[CacheConfig] = None,
         quant_config: Optional[QuantizationConfig] = None,
     ) -> None:
-        super().__init__()
+        nn.Module.__init__(self)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpm3.py:192]: This is a bit hacky. Perhaps it would be better to have a base class with the common methods and have both MiniCPM and MiniCPM-3 inherit from it.

Comment (SUDA-HLT-ywfang) [reply vllm/model_executor/models/minicpm3.py]: Thank you for your advice. I have optimized the structure of my implementation. Please review and let me known if you have any further questions.

Comment (DarkLight1337) [reply vllm/model_executor/models/minicpm3.py]: It looks better now, thanks!

File: vllm/model_executor/models/minicpm3.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                              quant_config=quant_config)
+
+    def forward(
+        self,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        kv_cache: torch.Tensor,
+        attn_metadata: AttentionMetadata,
+    ) -> torch.Tensor:
+        q = self.q_a_proj(hidden_states)[0]
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpm3.py:130]: To make the code a bit clearer, can you use tuple unpacking on the output of the projection layers? e.g.

```
q, _ = self.q_a_proj(hidden_states)
```

The current expression may be confused with slicing a tensor.

**Review by DarkLight1337 — COMMENTED** [review]
Overall looks good, mainly need to add a test as I mentioned earlier.

**Comment by SUDA-HLT-ywfang** [comment]
> Also, it would be great if you could add a test to show that the model matches the HF output.

Could you give an example please? Should I test on one case or an entire benchmark? And where should I add it?

**Comment by DarkLight1337** [comment]
We usually just test that the output of the model on some example prompts are consistent with those of HF. For example, you can take a look at `tests/models/test_models.py`.

**Comment by DarkLight1337** [comment]
Update: The new locations of the model tests have been adjusted in #7820, so please merge from main.

**Comment by SUDA-HLT-ywfang** [comment]
I have added a model test and run testing on 1xA100. The output of vllm on example prompts are consistent with those of HF.

File: vllm/model_executor/models/minicpm3.py (hunk 2 above)
Comment (SUDA-HLT-ywfang) [reply vllm/model_executor/models/minicpm3.py]: ok. I have fixed it in my latest commit.

**Review by DarkLight1337 — APPROVED** [review]
I've ran the test locally and got it to pass. Thanks for your hard work!
